### ISSUE TITLE
Using Vector db such as langchain for full-text search across the whole library

### ISSUE THREAD

**Issue by 00sapo** [MuiseDestiny/zotero-gpt#264]
Hello,
As of now, when asking questions with multiple papers as context, the plugin selects the abstract and put them all in the same message. This often results in a token number exceeded error. Moreover, the context is only limited to the abstracts.

The proper way to deal with this is to use a vector-db such as langchain js. Each time a new paper is added, it should also be added to the vector-db. Then, when the use asks for something, the Vector-db is queried and the retrieved top-k excerpts are used to contextualize the question to chatgpt.

I never built a zotero addon, but if you agree with this idea, I would be happy to help.

**Comment by zfb132** [comment]
@Edilson-R If you are still interested in this answer, you can check out the following content
> As of now, when asking questions with multiple papers as context, the plugin selects the abstract and put them all in the same message. This often results in a token number exceeded error. Moreover, the context is only limited to the abstracts.

Yes. It's a problem. If you select one or multiple articles in the following interface and then invoke GPT to ask a question. 
Indeed, only some metadata will be sent to OpenAI, not the full-text content.
![279556468-0dac8fac-c257-456b-be3b-d13ec96c2f4a](https://github.com/MuiseDestiny/zotero-gpt/assets/18099238/1ae84bc1-21bd-4155-80b3-0309da9d1e69)

> The proper way to deal with this is to use a vector-db such as langchain js. Each time a new paper is added, it should also be added to the vector-db. Then, when the use asks for something, the Vector-db is queried and the retrieved top-k excerpts are used to contextualize the question to chatgpt.

As far as I know, zotero-gpt is using the `langchain.js` and [OpenAIEmbeddings](https://js.langchain.com/docs/modules/data_connection/text_embedding/integrations/openai) when processing the full text. And the vectors for each paper will also be saved in local disk for future use.
https://github.com/MuiseDestiny/zotero-gpt/blob/e80732367f6804f7c36a8403a1ff026403e6eeee/src/modules/Meet/OpenAI.ts#L48-L150 

Therefore, this means that the plugin supports extracting the full text content, compressing it into vectors, and then sending it to GPT for answers. When you invoke GPT in the following interface (the opened PDF reading interface), you can use this feature. And you can pay attention to the pop-up box in the bottom right corner, which will display the work the plugin is currently performing in real time.
![image](https://github.com/MuiseDestiny/zotero-gpt/assets/18099238/e0365262-31de-40df-ab27-0048d298cf80)

> I never built a zotero addon, but if you agree with this idea, I would be happy to help.

Although this feature has already been implemented in the current version of the plugin, we are still open to any additional suggestions or improvements you might have. Thank you very much for your enthusiasm.

My tests were conducted based on the following versions:
- **OS**: Win 11 22H2 22621.2506
- **Zotero version**: 7.0.0-beta.47+f012a348a (64-bit)
- **Plugin version**: 0.3.2

**Comment by 00sapo** [comment]
Hey, thank you very much for your reply! It's nice to see that zotero-gpt is already using a vector-db. 

Do mean that the last pre-release 3.0 can use full text on multiple documents? Otherwise I would be happy to help for such a feature. I would also like to have embeddings for html and notes, beyond pdfs, because I use zotero as a personal note system.

The reason why I thought it wasn't using embeddings is that the last stable release on zotero 6 often goes out of tokens because the number of selected papers is to high (no excerpt extraction there) and the same happens when a note is selected, even though in that case it can see the "full-text" of the note as it is returned by zotero in the metadata.

**Comment by zfb132** [comment]
> Do mean that the last pre-release 3.0 can use full text on multiple documents?

If you're using GPT in the Zotero main window, it will only use the metadata as input.
If you're using GPT in the PDF reading interface, it will only use the current PDF file content as input.

> The reason why I thought it wasn't using embeddings is that the last stable release on zotero 6 often goes out of tokens because the number of selected papers is to high (no excerpt extraction there) and the same happens when a note is selected, even though in that case it can see the "full-text" of the note as it is returned by zotero in the metadata.

If you select one or multiple papers in the main window (not the PDF reading interface), the plugin will get the metadata, use embeddings, and use chat completions. For example:
1. You select 23 papers and input your question
2. The plugin will extract the metadata for the 1st~10th (`embeddingBatchNum`) papers and send the string to OpenAIEmbeddings
3. The same operations to 11th~20th
4. The same operations to 21st~23th
5. Select the top 6 (`relatedNumber`) relevant papers' metadata as input for GPT

**Comment by 00sapo** [comment]
Thanks. I still don't understand if you're talking about zotero 7 with pre-release version or zotero 6 with stable release. What I'm seeing in zotero 6 with the stable release is that from the main window with more than 10 papers selected or with a long note selected, it goes out of tokens, as if there was no excerpt selection from the input metadata.

**Comment by zfb132** [comment]
Zotero 6 is outdated, so the explanation is based on Zotero 7.
> OS: Win 11 22H2 22621.2506
Zotero version: 7.0.0-beta.47+f012a348a (64-bit)
Plugin version: 0.3.2

**Comment by 00sapo** [comment]
zotero 6 is the latest stable release zotero-gpt last release is still compatible with it. Zotero 6 is not outdated for most of Linux distro, because it's the latest stable release.

https://www.zotero.org/download/

**Comment by zfb132** [comment]
You're right, but upgrading from z6 to z7 is not smooth. It will take more time and energy to be compatible with z6. If we plan to continue support z6, the priority will still be lower than fix bug or develop new features on z7

**Comment by 00sapo** [comment]
Ok, so basically, I should check pre-release with zotero 7. I'll try it out and see how to introduce full-text multi-document embeddings

**Comment by zfb132** [comment]
Yes, you can. You should remember to back up all your data and config before upgrading in case of a critical error. Some plugins may not work on z7.

**Comment by Einsteinerly** [comment]
I just found that when using embeddings libraries, asking the same question on different PDFs seems to yield similar answers. Is this because all embeddings are stored in the same file?
